doc: Add Kubeflow tech review snapshot (#2180)
cncf/toc: projects/kubeflow/tech-review/2026-06-18.md

with additional integration and management tools. Kubeflow AI reference platform deploys the
comprehensive toolkit for the entire AI lifecycle. The Kubeflow AI reference platform can be
installed via [Packaged Distributions](https://www.kubeflow.org/docs/started/installing-kubeflow/#packaged-distributions)
or [Kubeflow Manifests](https://www.kubeflow.org/docs/started/installing-kubeflow/#kubeflow-manifests).

## Day 0 - Planning Phase

### Scope

#### Describe the roadmap process

Kubeflow projects are managed by community members that are part of working groups.
Each working group defines and agrees on the features for each release. The release cadence for
each working group varies according to community agreement among the working groups.

Then, all Kubeflow projects have individual roadmap files in the Git repository defining each
release and available to the public. This ensures we have a standard structure for each proposed
feature, auditing, versioning, and transparency, since it is recorded along with the history in the Git repo.

For more information, check ROADMAP for each Kubeflow Project:

> 

Reviewer: @christian-heusel Please add the Kubeflow Notebooks ROADMAP: https://github.com/kubeflow/notebooks/blob/main/ROADMAP.md

Author: Already added yesterday, see https://github.com/cncf/toc/pull/2180#discussion_r3437273601

Reviewer: oh, sorry, missed it.

> - [Kubeflow Spark Operator](https://github.com/kubeflow/spark-operator/blob/master/ROADMAP.md)

Reviewer 2: not blocking - worth noting that a number of these are dated by year and in those cases appear to be out-of-date.

Author: Good catch, I have opened issues for the ones that have been obviously out of date:
- https://github.com/kubeflow/spark-operator/issues/3023
- https://github.com/kubeflow/katib/issues/2681
- https://github.com/kubeflow/pipelines/issues/13613

- [Kubeflow Trainer](https://github.com/kubeflow/trainer/blob/master/ROADMAP.md)
- [Kubeflow Katib](https://github.com/kubeflow/katib/blob/master/ROADMAP.md)
- [Kubeflow Hub](https://github.com/kubeflow/hub/blob/main/ROADMAP.md)

> - [Kubeflow Pipelines](https://github.com/kubeflow/pipelines/blob/master/ROADMAP.md)

Reviewer 3: I am going through the graduation application and currently reviewing the ROADMAPs section, and hence came back to the GTR. kubeflow/notebooks is missing here.

Author: Currently in review in https://github.com/kubeflow/notebooks/pull/1188 as raised by @franciscojavierarceo 😄 I'll add a link to it and resolve this todo once the PR is merged! 🤗 

/hold

- [Kubeflow Notebooks](https://github.com/kubeflow/notebooks/blob/main/ROADMAP.md)

Community-wide changes are proposed as [Kubeflow Enhancement proposals (KEPs)](https://github.com/kubeflow/community/tree/master/proposals)

> in the `kubeflow/community` repository or in the [Kubeflow sub-projects KEPs](https://github.com/kubeflow/trainer/tree/master/docs/proposals).

Reviewer 2: 404 link

#### Describe the target persona or user(s) for the project

1. Users: Data Scientists, ML Engineer, AI Practitioners, Data Engineers, AI Practitioners.
2. Operators: MLOps Engineers, AIOps engineers, Platform Engineers, AI Platform Engineers.
3. Vendors: Vendors and projects building Kubernetes based AI Platform products.

#### Explain the primary use case for the project. What additional use cases are supported by the project

The goal of Kubeflow is to run Cloud Native AI workloads for every stage in AI lifecycle. By
using Kubeflow projects users can develop and deploy AI applications.

The primary use-cases include:

- Large-scale data processing and feature engineering.
- Distributed pre-training of foundation models.
- Post-training and fine-tuning of LLMs.
- Hyperparameter optimization and model tuning.
- LLM inference and multi-host serving.

Additional Use Cases:

- End-to-End GenAI pipeline building.
- Interactive AI development.
- Multiple users / projects with hard multi-tenancy on the same cluster.

#### Explain which use cases have been identified as unsupported by the project

> As Kubeflow is composed of multiple projects, each working group makes its own determinations as t

Reviewer 2: typo - perhaps:

```suggestion
As Kubeflow is composed of multiple projects, each working group makes its own determinations as to
```

what will be excluded from them. However we have an overarching theme and governance structure
(Steering Committee) that has identified the following areas as not being a priority for all projects:

- The projects are deployed in any Kubernetes (each release will specify tested versions),
  regardless of the underlying infrastructure, independently through Kubernetes manifests leveraging
  Kustomize and/or Helm Charts. However, the project doesn’t provide an implementation to be deployed

>   on infrastructure besides Kubernetes. - We do not officially enforce a deployment method or distribution.

Reviewer 2: Is this last bullet meant to be included in the surrounding bullet?

Reviewer 4: This looks like it was meant to be an em-dash, not a bullet-point delimiter.

- Kubeflow doesn’t provide a GitOps implementation, however Kubeflow manifests can be integrated
  into a GitOps solution. For example, Platform Engineers can create an ArgoCD Application (CRD)

>   to install and configure Kubeflow projects. by providing Kubeflow individual project manifests,

Reviewer 2: ```suggestion
  to install and configure Kubeflow projects by providing Kubeflow individual project manifests,
```

customizations on top of the Kubeflow platform or choose to build distributions to help other
organizations adopt a customized platform based on Kubeflow projects.

As Kubeflow maintains flexibility, organizations can choose their own path according to their needs.

We encourage adopters to be part of the adopters list in [GitHub](https://github.com/kubeflow/community/blob/master/ADOPTERS.md).

Examples of organizations that use Kubeflow:

- AWS
- Red Hat
- Capital One
- CERN
- Google
- Alibaba Cloud
- Bloomberg
- IBM
- Cisco
- Huawei
- Microsoft
- Tencent
- DHL Data & Analytics
- Telia
- Roblox
- Toyota
- PepsiCo
- Volvo
- and others…

#### Describe any completed end user research and link to any reports.

We regularly undertake user research about Kubeflow and its users. Research could be done by
working groups, during events, or by conducting at least one study annually. We ensure these surveys
are visible to the public and shared with the community.

Here are some results from previous years.

- [Kubeflow 2025 Survey](https://docs.google.com/forms/d/11cSe5vmGLrGekJISHBMfjVh_97WFGuhcvGnd0l5aNLg/edit#responses)

> - [2025:UX designers supporting Model Registry conducted a series of user sessions to understand preferred interaction patterns (link)](https://docs.google.com/forms/d/11cSe5vmGLrGekJISHBMfjVh_97WFGuhcvGnd0l5aNLg/edit#responses)

Reviewer 2: not blocking - These two links reference the survey itself and I cannot see the responses.

Author: @andreyvelich @jaiakash I also added https://blog.kubeflow.org/kubeflow-sdk-user-survey-insights/ here (https://github.com/kubeflow/blog/pull/198)

- [Kubeflow Survey 2024](https://github.com/kubeflow/community/issues/708)
- [Kubeflow Survey 2023](https://blog.kubeflow.org/kubeflow-user-survey-2023/)
- [Kubeflow Survey 2022](https://blog.kubeflow.org/kubeflow-user-survey-2022/)

### Usability

#### How should your target personas interact with your project

AI Practitioner - Kubeflow SDK, Kubeflow UIs

Platform Admins - Operator guides, installing and configuring any Kubeflow projects via Helm charts
or Kustomize manifests predefined and available in the Kubeflow documentation from the command line.

- [Getting Started guide](https://www.kubeflow.org/docs/started/introduction/).
- [How to install Kubeflow projects](https://www.kubeflow.org/docs/started/installing-kubeflow/).
- [How to contribute to Kubeflow](https://www.kubeflow.org/docs/about/contributing/).

#### Describe the user experience (UX) and user interface (UI) of the project

> Kubeflow user experience in each project is a collection of projects, the user experience for the

Reviewer 2: > Kubeflow user experience in each project is a collection of projects

Is this accurate?

Author: I tried to clarify this to the best of my ability in the latest revision 😊

projects are each with their own [interfaces, APIs and SDKs](https://www.kubeflow.org/docs/started/architecture/#kubeflow-interfaces).

##### Describing User Experience through SDK

We are working on a unified [Kubeflow SDK](https://github.com/kubeflow/sdk) that gives
AI practitioners Python-native experience to interact with Kubeflow APIs.

Through Kubeflow SDK, users will be able to interact with the different projects, increasing user experience and reducing complexity.

<div style="text-align: center;">
  <img

>     src="https://raw.githubusercontent.com/kubeflow/sdk/main/docs/images/persona_diagram.svg"

Reviewer 2: 404 link

Kubeflow projects support multi-tenancy, which allows organizations to isolate workloads in a
Kubernetes cluster, which can be used with Network Policies to restrict isolation from different
components by ports or application name. Kubeflow projects support custom service accounts,
allowing organizations to control how Kubeflow projects interact with the cluster by leveraging RBAC.
Further, Platform Administrators can decide which users in Kubernetes will have access to the
Kubeflow projects by creating different users and RBAC policies. Kubeflow Notebooks also supports
profiles, which allows simplifying the permission management.

In terms of data management, data sovereignty is managed by those who deploy or package the projects.

#### Describe any compliance requirements addressed by the project

Kubeflow projects are extensible which allows users to fit their internal compliance requirements.
As a result, specific common compliance frameworks (SOC-2, GDPR, etc.) are the responsibility of
end users and vendors. However, we aim to provide a strong foundation through reference architectures
similar things from which to build on.

> #### Describe the project’s High Availability requirements

Reviewer 2: I don't think this reference is accurate anymore - but regardless this may be an area of improvement for future iterations. I imagine there are more specific requirements to consider for each project and high availability than simply adjusting replicas?

  by using tools and GitHub workflows that detect sensitive data in PRs.

- **License compliance:** Kubeflow uses
  [the FOSSA product](https://app.fossa.com/reports/bb8e2d41-254a-4af3-9044-e7f484c34dd1)
  provided by the CNCF to scan for license compliance on a regular and ongoing basis.

## Day 1 - Installation and Deployment Phase

### Project Installation and Configuration

#### Describe what project installation and configuration look like

Each project has its [own standalone installation guide](https://www.kubeflow.org/docs/started/installing-kubeflow/#standalone-kubeflow-components).

Kubeflow projects are packaged by [distributions and vendors](https://www.kubeflow.org/docs/started/installing-kubeflow/#kubeflow-platform)
into platform products which can be used to install our tools.

We provide optional manifests to deploy all Kubeflow projects as
[Kubeflow AI reference platform](https://github.com/kubeflow/manifests/blob/master/README.md).

### Project Enablement and Rollback

#### How can this project be enabled or disabled in a live cluster? Please describe any downtime required of the control plane or nodes

> Users can set the replica count to 0 in the Kubeflow projects deployment. Existing AI workloads

Reviewer 2: Seems functional but maybe not a pragmatic choice for handling "disabling" - Does this offer any benefits over more declarative abstractions (why not remove on disable versus scaling to zero?)

are evaluated and upgraded on every release. We support
[Kubernetes 1.31+ and test on 1.32+](https://github.com/kubeflow/manifests/blob/master/tests/install_KinD_create_KinD_cluster_install_kustomize.sh)

#### How the project handles rollback procedures

Some projects support leader election and HA to make sure that long-running workloads are complete.
Users can scale up or down replicas of controllers during rollback.

#### How can a rollout or rollback fail? Describe any impact to already running workloads

Newer revision will not be active for many reasons including insufficient cluster resources for the
newer spec, invalid configuration spec. Traffic will not be switched to the new unless the newer
revision is ready to accept traffic. Hence, the already running workloads will not be affected
in any case if rollout fails.

#### Describe any specific metrics that should inform a rollback

Kubeflow projects CRDs expose status that inform about activity of AI workloads. Additionally,
controllers are exposed various Prometheus metrics to indicate workload status.

#### Explain how upgrades and rollbacks were tested and how the upgrade->downgrade->upgrade path was tested

> Currently, it’s being manually tested by users, but automated tests are work in progress.

Reviewer 2: References would be great.

Author: Added in my latest changes ✅

API versions of CRDs.

## Day 2 - Day-to-Day Operations Phase

### Scalability/Reliability

#### Describe how the project increases the size or count of existing API objects

In general, the API object count complexity is linear in the number of users and workloads. Users
can list deployed Kubeflow CRDs across all namespaces.

For example, to list all TrainJob from Kubeflow Trainer run:

```sh
kubectl get trainjob --all-namespaces
```

The core controllers and resources are constant (beyond replicating specific controllers),
see [the full kustomize build here](https://github.com/kubeflow/manifests?tab=readme-ov-file#install-with-a-single-command).

#### Describe how the project defines Service Level Objectives (SLOs) and Service Level Indicators (SLIs)

Kubeflow currently doesn’t provide the SLOs and SLIs, however we follow
[the Kubernetes SLOs](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md)

#### Describe any operations that will increase in time covered by existing SLIs/SLOs

As above.

#### Describe the increase in resource usage in any components as a result of enabling this project, to include CPU, Memory, Storage, Throughput

> Resources requirements for Kubeflow projects [are set here](https://github.com/kubeflow/manifests/pull/3091#issuecomment-3016609243).

Reviewer 2: The table highlights the max usages (as I understood it) whereas the comment also captured actual use. Maye be worth a distinction between "here is the minimum resource usage to expect being used versus the maximum for planning on both ends.

Kubeflow uses the Kubernetes resilience pattern to manage controllers with HA. If one replica fails, another part of the control plane takes responsibility to orchestrate workloads.

### Observability Requirements

#### Describe the signals the project is using or producing, including logs, metrics, profiles and traces. Please include supported formats, recommended configurations and data storage

Kubeflow controllers expose Prometheus metrics to report workload status. Controllers also expose
logs and status for platform admins to ensure stability.

#### Describe how the project captures audit logging

Platform admins can leverage [Kubernetes audit](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/)
for Kubeflow projects.

#### Describe any dashboards the project uses or implements as well as any dashboard requirements

[Kubeflow Dashboard requirements](https://www.kubeflow.org/docs/components/central-dash/overview/).

#### Describe how the project surfaces project resource requirements for adopters to monitor cloud and infrastructure costs, e.g. FinOps That must happen on the Kubernetes namespace level

> Users are recommended to use third-party tools like Kubecost to measure cloud and infrastructure

Reviewer 2: Is this an implicit or explicit recommendation (I did not see a docs reference). I think it's fine to defer this to other processes and metrics.

cost of running Kubeflow projects.

#### Which parameters is the project covering to ensure the health of the application/service and its workloads

Most project use Kubernetes [Liveness and Readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/).

For example [Kubeflow Trainer controller manager](https://github.com/kubeflow/trainer/blob/master/manifests/base/manager/manager.yaml#L30-L43).

#### How can an operator determine if the project is in use by workloads

> - Check the Pods in `kubeflow-profil`e labeled namespaces.

Reviewer 2: ```suggestion
- Check the Pods in `kubeflow-profile` labeled namespaces.
```

- Check the CRDs in user’s namespaces
- Check the Kubeflow Dashboard resources.

> #### How can someone using this project know that it is working for his instance

Reviewer 2: ```suggestion
#### How can someone using this project know that it is working for their instance
```

#### How does the project incorporate and consider source composition analysis as part of its development and security hygiene? Describe how this source composition analysis (SCA) is tracked

Various static and dynamic code analysis tools are enforced as described above.

#### Describe how the project implements changes based on source composition analysis (SCA) and the timescale

Various static and dynamic code analysis tools are enforced as described above.

### Troubleshooting

#### How does this project recover if a key component or feature becomes unavailable

Specifics depend on the Kubeflow Project (KFP, Katib, etc.). Kubeflow projects are cloud and
Kubernetes native apps, so fault tolerance is strongly tied to the health of the underlying cluster.
In practice user workloads sometimes have a retry policy and the Kubeflow core services should
just recover automatically.

More details are described in the previous sections.

#### Describe the known failure modes

Each Kubeflow project handles failure modes differently beyond native Kubernetes fault tolerance.
Many of them are configured at the application level in user code.

> ### Security

Reviewer 2: ```suggestion

### Compliance

  * What steps does the project take to ensure that all third-party code and components have correct and complete attribution and license notices?
  * Describe how the project ensures alignment with CNCF [recommendations](https://github.com/cncf/foundation/blob/main/policies-guidance/recommendations-for-attribution.md) for attribution notices.
    <!--Note that each question describes a use case covered by the referenced policy document.-->
    * How are notices managed for third-party code incorporated directly into the project's source files?
    * How are notices retained for unmodified third-party components included within the project's repository?
    * How are notices for all dependencies obtained at build time included in the project's distributed build artifacts (e.g. compiled binaries, container images)?

### Security
```

The compliance section is missing from this document.

Author: I was able to answer the first question with a pointer to FOSSA; 

However for the others I'll need help Could you provide some guidance on this topic @juliusvonkohout @chasecadet @andreyvelich @franciscojavierarceo @thesuperzapper (as current Kubeflow Steering Committee members).

Reviewer: I think, FOSSA should cover all of those questions. Isn't?

Author: Awesome, I have amended the section to answer to all of these questions 🤗

#### How does the project incorporate and consider source composition analysis as part of its development and security hygiene? Describe how this source composition analysis (SCA) is tracked

Various static and dynamic code analysis tools are enforced as described above.

#### Describe how the project implements changes based on source composition analysis (SCA) and the timescale

Various static and dynamic code analysis tools are enforced as described above.

### Troubleshooting

#### How does this project recover if a key component or feature becomes unavailable

Specifics depend on the Kubeflow Project (KFP, Katib, etc.). Kubeflow projects are cloud and
Kubernetes native apps, so fault tolerance is strongly tied to the health of the underlying cluster.
In practice user workloads sometimes have a retry policy and the Kubeflow core services should
just recover automatically.

More details are described in the previous sections.

#### Describe the known failure modes

Each Kubeflow project handles failure modes differently beyond native Kubernetes fault tolerance.
Many of them are configured at the application level in user code.

> ### Compliance

Reviewer 2: Consider formatting of the questions under this section when addressed.
